File: core/src/main/java/net/dontdrinkandroot/wicket/model/ChainedModel.java
package net.dontdrinkandroot.wicket.model;

import org.apache.wicket.model.IModel;

/**
 * An {@link IModel} that resolves its object based on another parent Model. It therefore does not have to hold any
 * data itsself but rather knows how to transform the data of the parent model.
 *
 * @param <P> Type of the parent Model.
 * @param <T> Type of the resolved model object.
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public interface ChainedModel<P, T> extends IModel<T>
{
    IModel<? extends P> getParent();

    P getParentObject();
}

File: core/src/main/kotlin/net/dontdrinkandroot/wicket/model/ChainedModel.kt
package net.dontdrinkandroot.wicket.model

import org.apache.wicket.injection.Injector
import org.apache.wicket.model.IModel

abstract class AbstractChainedModel<P, T>(protected val parent: IModel<out P?>) : IModel<T> {

    final override fun getObject(): T? = this.getValue(parent.getObject())

    abstract fun getValue(parentValue: P?): T?

    override fun detach() {
        this.parent.detach()
    }
}

abstract class AbstractChainedInjectedModel<P, T>(parent: IModel<out P?>) :
    AbstractChainedModel<P, T>(parent) {

    init {
        Injector.get().inject(this)
    }
}

fun <P, T> IModel<P>.chain(getChain: (P?) -> T, setChain: (P?, T?) -> Unit = { _: P?, _: T? -> }): IModel<T> =
    object : AbstractChainedModel<P, T>(this) {
        override fun getValue(parentValue: P?): T = getChain(parentValue)
        override fun setObject(value: T?) = setChain(this.parent.getObject(), value)
    }
